src/trusted/service_runtime/sel_ldr_standard.c

```diff
@@ -904,6 +874,215 @@ int NaClCreateMainThread(struct NaClApp     *nap,
+   * overflowed.
+   *
+   * NB: the underlying OS would have limited the amount of space used
+   * for argv and envv -- on linux, it is ARG_MAX, or 128KB -- and
+   * hence the overflow check is for obvious auditability rather than
+   * for correctness.
+   */
+  auxv_entries = 1;
+  if (0 != nap_child->user_entry_pt) {
+    auxv_entries++;
+  }
+  ptr_tbl_size = (((NACL_STACK_GETS_ARG ? 1 : 0) +
+                   (3 + argc + 1 + envc + 1 + auxv_entries * 2)) *
+                  sizeof(uint32_t));
+
+  if (SIZE_T_MAX - size < ptr_tbl_size) {
+    NaClLog(LOG_WARNING,
+            "NaClCreateMainThread: ptr_tbl_size cause size of"
+            " argv / environment copy to overflow!?!\n");
+    retval = 0;
+    goto cleanup;
+  }
+  size += ptr_tbl_size;
+
+  size = (size + NACL_STACK_ALIGN_MASK) & ~NACL_STACK_ALIGN_MASK;
+
+  if (size > nap_child->stack_size) {
+    retval = 0;
+    goto cleanup;
+  }
+
+  /*
+   * Write strings and char * arrays to stack.
+   */
+  stack_ptr = NaClUserToSysAddrRange(nap_child,
+                                     NaClGetInitialStackTop(nap_child) - size,
+                                     size);
+  if (stack_ptr == kNaClBadAddress) {
+    retval = 0;
+    goto cleanup;
+  }
+
+  NaClLog(2, "setting stack to : %016"NACL_PRIxPTR"\n", stack_ptr);
+
+  VCHECK(!(stack_ptr & NACL_STACK_ALIGN_MASK),
+         ("stack_ptr not aligned: %016"NACL_PRIxPTR"\n", stack_ptr));
+
+  p = (uint32_t *)stack_ptr;
+  strp = (char *)stack_ptr + ptr_tbl_size;
+
+  /*
+   * For x86-32, we push an initial argument that is the address of
+   * the main argument block.  For other machines, this is passed
+   * in a register and that's set in NaClStartThreadInApp.
+   */
+  if (NACL_STACK_GETS_ARG) {
+    uint32_t *argloc = p++;
+    *argloc = (uint32_t) NaClSysToUser(nap_child, (uintptr_t) p);
+  }
+
+  *p++ = 0;  /* Cleanup function pointer, always NULL.  */
+  *p++ = envc;
+  *p++ = argc;
+
+  for (i = 0; i < argc && argv; ++i) {
+    *p++ = (uint32_t)NaClSysToUser(nap_child, (uintptr_t)strp);
+    NaClLog(2, "copying arg %d  %p -> %p\n",
+            i, (void *)argv[i], (void *)strp);
+    snprintf(strp, ARG_LIMIT, "%s", argv[i]);
+    strp += argv_len[i];
+  }
+  *p++ = 0;  /* argv[argc] is NULL.  */
+
+  for (i = 0; i < envc; ++i) {
+    *p++ = (uint32_t)NaClSysToUser(nap_child, (uintptr_t)strp);
+    NaClLog(2, "copying env %d  %p -> %p\n",
+            i, (void *)envv[i], (void *)strp);
+    snprintf(strp, ARG_LIMIT, "%s", envv[i]);
+    strp += envv_len[i];
+  }
+  *p++ = 0;  /* envp[envc] is NULL.  */
+
+  /* Push an auxv */
+  if (nap_child->user_entry_pt) {
+    *p++ = AT_ENTRY;
+    *p++ = (uint32_t)nap_child->user_entry_pt;
+  }
+  *p++ = AT_NULL;
+  *p++ = 0;
+
+  CHECK((char *)p == (char *)stack_ptr + ptr_tbl_size);
+
+  /* now actually spawn the thread */
+  NaClXMutexLock(&nap_child->mu);
+  nap_child->running = 1;
+  NaClXMutexUnlock(&nap_child->mu);
```

Comment:
Can you explain why the critical section here is only for the switching of the running flag for the `nap_child`?

Reply:
We need a lock for two reasons:

1) To act as a memory barrier and synchronization point for when the parent NaClApp and functions like NaClReportExitStatus() or try to query thread running state.

2) To block strange spurious thread spawns when running on a host with glibc 2.28 or later.

I am still trying to figure out the reason behind #2, but I've had no luck yet. C11 threads *were* added in glibc 2.28 which may be related, but haven't found anything concrete so far.

Reply:
For what it's worth, I did copy this lock straight from the non-fork thread constructor function, so it is possible this may not be absolutely needed here.

I mostly looked at other code which read this variable and the lock seemed sane, but I'll do some testing to verify it. Good catch!